cpp/src/lists/contains.cu

```diff
@@ -44,296 +41,243 @@
 
 #include <type_traits>
 
-namespace cudf {
-namespace lists {
+namespace cudf::lists {
 
 namespace {
 
-auto constexpr absent_index = size_type{-1};
-
-auto get_search_keys_device_iterable_view(cudf::column_view const& search_keys,
-                                          rmm::cuda_stream_view stream)
-{
-  return column_device_view::create(search_keys, stream);
-}
-
-auto get_search_keys_device_iterable_view(cudf::scalar const& search_key, rmm::cuda_stream_view)
-{
-  return &search_key;
-}
+/**
+ * @brief A sentinel value used for marking that a given key has not been found in the search list.
+ */
+auto constexpr __device__ NOT_FOUND_SENTINEL = size_type{-1};
 
-template <typename ElementType, duplicate_find_option find_option>
-auto __device__ find_begin(list_device_view const& list)
+template <typename Type>
+auto constexpr is_supported_non_nested_type()
 {
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return list.pair_rep_begin<ElementType>();
-  } else {
-    return thrust::make_reverse_iterator(list.pair_rep_end<ElementType>());
-  }
+  return cudf::is_fixed_width<Type>() || std::is_same_v<Type, cudf::string_view>;
 }
 
-template <typename ElementType, duplicate_find_option find_option>
-auto __device__ find_end(list_device_view const& list)
+/**
+ * @brief Return a pair of index iterators {begin, end} to loop through elements within a list.
+ *
+ * Depending on the value of `find_first`, a pair of forward or reverse iterators will be returned,
+ * allowing to loop through elements in the list in first-to-last or last-to-first order.
+ *
+ * Note that the element indices always restart to `0` at the first position in each list.
+ *
+ * @tparam find_first A boolean value indicating whether we want to find the first or last
+ *         appearance of a given key in the list.
+ * @param size The number of elements in the list.
+ * @return A pair of {begin, end} iterators to iterate through the range `[0, size)` by forward or
+ *         reverse order.
+ */
+template <bool find_first>
+auto __device__ element_index_pair_iter(size_type const size)
 {
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return list.pair_rep_end<ElementType>();
+  if constexpr (find_first) {
+    return thrust::pair(thrust::make_counting_iterator<size_type>(0),
```

Comment:
CTAD! Nice.

Reply:
This is interesting. I wonder why this error did not manifest in the old code.

Reply:
> Could this be due to anonymous space?

The anonymous namespace sounds like a likely culprit to me.

Reply:
I tried it in Godbolt too, even anonymous namespace. could not replicate it in 11.5, 11.2.

Reply:
Thanks for that @PointKernel. I didn't realize that guidance applied to variables as well as functions, but I agree it's reasonable to prefer the explicit nature of `__device__`.